src/timeline/TimelineModel.cpp

```diff
@@ -1934,11 +1934,298 @@ TimelineModel::formatPowerLevelEvent(const QString &id)
+                .arg(sender_name,
+                     QString::number(prevEvent->content.events_default),
+                     QString::number(event->content.events_default)));
+        } else if ((event->content.events_default < prevEvent->content.events_default) &&
+                   (event->content.events_default < default_powerlevel) &&
+                   (prevEvent->content.events_default > default_powerlevel)) {
+            resultingMessage.append(
+              tr("%1 has changed the room's events_default powerlevel from %2 to %3. New "
+                 "users can now send events that are not otherwise restricted.")
+                .arg(sender_name,
+                     QString::number(prevEvent->content.events_default),
+                     QString::number(event->content.events_default)));
+        } else {
+            resultingMessage.append(
+              tr("%1 has changed the room's events_default powerlevel from %2 to %3.")
+                .arg(sender_name,
+                     QString::number(prevEvent->content.events_default),
+                     QString::number(event->content.events_default)));
+        }
+    }
+
+    // Compare if a Powerlevel of a user changed
+    for (auto const &[mxid, powerlevel] : event->content.users) {
+        auto nameOfChangedUser = utils::replaceEmoji(displayName(QString::fromStdString(mxid)));
+        if (prevEvent->content.user_level(mxid) != powerlevel) {
+            if (powerlevel == administrator_power_level) {
+                resultingMessage.append(tr("%1 has made %2 an administrator of this room.")
+                                          .arg(sender_name, nameOfChangedUser));
+            } else if (powerlevel == moderator_power_level &&
+                       powerlevel > prevEvent->content.user_level(mxid)) {
+                resultingMessage.append(tr("%1 has made %2 a moderator of this room.")
+                                          .arg(sender_name, nameOfChangedUser));
+            } else if (powerlevel == moderator_power_level &&
+                       powerlevel < prevEvent->content.user_level(mxid)) {
+                resultingMessage.append(tr("%1 has downgraded %2 to moderator of this room.")
+                                          .arg(sender_name, nameOfChangedUser));
+            } else {
+                resultingMessage.append(tr("%1 has changed the powerlevel of %2 from %3 to %4.")
+                                          .arg(sender_name,
+                                               nameOfChangedUser,
+                                               QString::number(prevEvent->content.user_level(mxid)),
+                                               QString::number(powerlevel)));
+            }
+        }
+    }
+
+    // Handle added/removed/changed event type
+    for (auto const &[event_type, powerlevel] : event->content.events) {
+        if (prevEvent->content.event_level(event_type) != powerlevel) {
```

Comment:
what if this is a state event?

Reply:
Good question and I dont have an immediate idea how to do it then :D

Reply:
actually nvm. I do